Generate changelog with towncrier

pquentin opened the pull request:
I also added the changelog to the docs, inspired by pip (even though they use a plugin to reformat their changes).

Here's what this looks like for now: https://urllib3--2330.org.readthedocs.build/en/2330/changelog.html

![Screenshot 2021-07-16 at 09 56 29](https://user-images.githubusercontent.com/42327/125899231-17c4923f-bfc5-4756-badf-8d5fb122e745.png)

The goal of this PR is to reproduce the existing changelog with towncrier, and make sure it looks correctly in docs. In follow-up PRs we can go through the 2.0 commits and add them to newsfragments.

Closes #2085

pquentin commented on src/urllib3/response.py:
```diff
@@ -906,6 +906,10 @@ def url(self) -> Optional[str]:
+    @url.setter
+    def url(self, url: str) -> None:
+        self._request_url = url
+
```
This is a hacky workaround to allow Sphinx to set the URL:

https://github.com/sphinx-doc/sphinx/blob/564439e66d81693edd08fa5973a14151af976ca8/sphinx/ext/intersphinx.py#L118-L120

"hacky" because it does not work when there are redirects. (We should first agree that whether we want to allow this or not. I think we do.)

sethmlarson replied:
I think allowing this is fine.

sigmavirus24 reviewed (approved):
I love this. Towncrier's a pleasure to use in Twine

pquentin commented:
@sigmavirus24 Thank you for your support, and thanks for mentioning twine! This gave me the good idea or renaming the obscure `newsfragment` directory into `changelog`.

sethmlarson reviewed (dismissed):
Woo, this is very nice. Thanks for tackling this Quentin! :tada:

Are we okay to merge now or is another test case required for setting `url`?

pquentin commented:
> Are we okay to merge now or is another test case required for setting url?

I'm not happy with the way this works in the `if self.retries is not None and self.retries.history:` case. But now that I look at the 1.26.x code, `url` was just not a thing in the past: we had `geturl()`. So maybe it's fine to say that Sphinx is at fault here and should not attach an `url` attribute? (They have a really fast release cycle, and I can probably get that fix in.)

bluetech reviewed (approved):
I've had a good experience with towncrier in pytest. It's nice for the PR submitter (can update the changelog without fear of conflicts), the release manager (doesn't need to create the changelog themselves), and for reviewers (the changelog is nice to have when reviewing).

I recommend adding a README file to the `changelog` directory explaining various things like
- The format of the filename (PR/issue number, category, rst)
- Which tense to use (present, past, etc.) with examples.
- Which categories (removal, feature, etc.) are possible and when to use them.

Even though it duplicates the contributing docs, it's very helpful for a quick reference, even for regular contributors.

One slight annoyance I've had is when submitting a PR without an issue, you don't know the PR ID before submitting it, so must do a push + immediate force push for the changelog name. Don't know if there's a fix for that, but it's pretty minor anyway.

sethmlarson reviewed (approved):
Very excited about this! I only have one question:

pquentin commented on src/urllib3/_version.py:
```diff
@@ -1,2 +1 @@
-# This file is protected via CODEOWNERS
```
It's not useful until we have automated releases. I added it back anyway as it wasn't related to that pull request.

pquentin commented:
Please take another look:

- I rebased to cleanly separate the towncrier and url changes. (Please don't squash.)
- I added a README in the changelog directory, using pytest as an inspiration.
- I addressed Seth's comment.
